Logdog ecs

webern opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**

Closes #1039

**Description of changes:**

Statically compile variant-specific logdog behavior by reading from two text files at build time. A symlink is used to point to the correct file for variant-specific commands.

**Testing done:**

Ran an ecs variant, got ecs logs. Ran a k8s variant, got k8s logs.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

webern commented on sources/logdog/static/aws-ecs-1:
```diff
@@ -0,0 +1,5 @@
+http get http://localhost:51678/v1/tasks
```
```suggestion
http ecs-tasks http://localhost:51678/v1/tasks
```

webern replied:
[afdc876](https://github.com/bottlerocket-os/bottlerocket/compare/6eb611d17506ab4aad98e5db4cd17b00bacd0839..afdc8765a9e97bb1d4436eda9b0c964fce24ab95)

tjkirch commented on sources/logdog/src/commands.rs:
```diff
@@ -0,0 +1,154 @@
+
+/// A logdog `Command` is in the format `mode filename [args...]`. `mode` specifies what type of
+/// command it is, e.g. `sh` for a shell command or `http` for an HTTP get request. `filename` is
+/// the name of the output file. `args` is any number of shell-like arguments. Putting this all
```
```suggestion
/// the name of the output file. `args` is any number of shell-like arguments, for `sh`, or a URL, for `http`. Putting this all
```
I guess "shell-like" doesn't apply to the http arguments...

webern replied:
[afdc876](https://github.com/bottlerocket-os/bottlerocket/compare/6eb611d17506ab4aad98e5db4cd17b00bacd0839..afdc8765a9e97bb1d4436eda9b0c964fce24ab95)

tjkirch commented on sources/logdog/Cargo.toml:
```diff
@@ -10,10 +10,12 @@ exclude = ["README.md"]
 
 [dependencies]
 flate2 = "1.0"
+reqwest = { version = "0.10.1", default-features = false, features = ["blocking"] }
```
Should we add `rustls-tls` to the feature list to support https URLs?

webern replied:
[afdc876](https://github.com/bottlerocket-os/bottlerocket/compare/6eb611d17506ab4aad98e5db4cd17b00bacd0839..afdc8765a9e97bb1d4436eda9b0c964fce24ab95)

tjkirch commented on sources/logdog/static/common:
```diff
@@ -0,0 +1,15 @@
+sh containerd-config containerd config dump
```
nit, because it's really a question about the existing command: do we need to do something to talk to the host-containerd specifically, rather than the containerd used for k8s?

webern replied:
I'm not sure, both of these commands have the exact same output in sudo sheltie:

```
containerd --address /run/host-containerd/containerd.sock config dump
containerd --address /run/containerd/containerd.sock config dump
```

Both of these report:

```
version = 2
root = "/var/lib/containerd"
state = "/run/containerd"
plugin_dir = ""
disabled_plugins = ["io.containerd.internal.v1.opt", "io.containerd.snapshotter.v1.aufs", "io.containerd.snapshotter.v1.devmapper", "io.containerd.snapshotter.v1.native", "io.containerd.snapshotter.v1.zfs"]
required_plugins = []
oom_score = 0
imports = ["/etc/containerd/config.toml"]

[grpc]
  address = "/run/dockershim.sock"
  tcp_address = ""
  tcp_tls_cert = ""
  tcp_tls_key = ""
  uid = 0
  gid = 0
  max_recv_message_size = 16777216
  max_send_message_size = 16777216
etc...
```

It seems like both of these report the non-host-containerd config (not sure though). I'm hoping we can avoid blocking on this.

samuelkarp replied:
You need `containerd --config /etc/host-containerd/config.toml config dump` to get the host-containerd config data.

webern replied:
[aa7d592](https://github.com/bottlerocket-os/bottlerocket/compare/60e5d3c8e65c58170e8f72d944a80be11ea098e0..aa7d59202dad5161a9dc8aff3995f1fbe0d4c706)

tjkirch commented on sources/logdog/static/aws-ecs-1:
```diff
@@ -0,0 +1,5 @@
+sh docker-daemon.json cat /etc/docker/daemon.json
+sh docker-info docker info
+sh ecs-agent-state.json cat /var/lib/ecs/data/ecs_agent_data.json
+sh ecs-config.json cat /etc/ecs/ecs.config.json
```
Maybe a `file` mode to handle the common case of copying files like this?

webern replied:
[aa7d592](https://github.com/bottlerocket-os/bottlerocket/compare/60e5d3c8e65c58170e8f72d944a80be11ea098e0..aa7d59202dad5161a9dc8aff3995f1fbe0d4c706)

tjkirch commented on sources/logdog/build.rs:
```diff
@@ -30,3 +39,36 @@ fn main() {
+        build. Valid values are the directories in models/src/variants/, for example 'aws-k8s-1.17'.", ENV_VARIANT);
+        process::exit(1);
+    });
+    if !PathBuf::from(format!("static/{}", variant)).is_file() {
```
I'm a little confused about this directory; let me suggest something...

We make `sources/logdog/etc` (or `conf` or something if we hate `etc`) where we keep our variant-specific files, which we'd name `logdog.VARIANT.conf`.  We have our writable directory that we mount in via the Dockerfile, and we put it at `soures/logdog/etc/current`, and we symlink `current/logdog.conf` -> `../logdog.VARIANT.conf`.  Then I think we don't need the copies in the Dockerfile, and (imo) it turns out a little simpler.  (And we could still have `sources/logdog/etc/logdog.common.conf` or similar.)

webern replied:
How's this?

```
├── .gitignore # (ignore conf/current/)
├── Cargo.toml
├── README.md
├── README.tpl
├── build.rs
├── conf
│   ├── aws-k8s.conf
│   ├── current
│   │   ├── logdog.common.conf -> ../logdog.common.conf
│   │   └── logdog.conf -> ../logdog.aws-k8s-1.17.conf
│   ├── logdog.aws-dev.conf
│   ├── logdog.aws-ecs-1.conf
│   ├── logdog.aws-k8s-1.15.conf -> aws-k8s.conf
│   ├── logdog.aws-k8s-1.16.conf -> aws-k8s.conf
│   ├── logdog.aws-k8s-1.17.conf -> aws-k8s.conf
│   └── logdog.common.conf
└── src
    ├── commands.rs
    ├── create_tarball.rs
    ├── error.rs
    └── main.rs
```

tjkirch replied:
Only logdog itself needs to read these files, so I don't think the common file needs to be linked into the writable directory, it'd just complicate build.rs.  Looks good though. :+1:

webern replied:
Oh yeah, good point, thanks.

webern replied:
[afdc876](https://github.com/bottlerocket-os/bottlerocket/compare/6eb611d17506ab4aad98e5db4cd17b00bacd0839..afdc8765a9e97bb1d4436eda9b0c964fce24ab95)

tjkirch commented on sources/logdog/build.rs:
```diff
@@ -1,11 +1,20 @@
+// variant we're building, and therefore which API model to use.
+const ENV_VARIANT: &str = "VARIANT";
 
 fn main() {
+    println!("cargo:rerun-if-env-changed={}", ENV_VARIANT);
+    symlink_variant();
+
     // Check for environment variable "SKIP_README". If it is set,
```
nit: can we do like `models` and put variant/readme code in separate functions, and just call those from main?

webern replied:
[aa7d592](https://github.com/bottlerocket-os/bottlerocket/compare/60e5d3c8e65c58170e8f72d944a80be11ea098e0..aa7d59202dad5161a9dc8aff3995f1fbe0d4c706)

tjkirch commented on Dockerfile:
```diff
@@ -79,6 +81,7 @@ USER builder
 RUN --mount=source=.cargo,target=/home/builder/.cargo \
     --mount=type=cache,target=/home/builder/.cache,from=cache,source=/cache \
     --mount=type=cache,target=/home/builder/rpmbuild/BUILD/sources/models/src/variant,from=variantcache,source=/variantcache \
+    --mount=type=cache,target=/home/builder/rpmbuild/BUILD/sources/logdog/static,from=variantcache,source=/variantcache/logdog \
```
I think we should consider doing `/variantcache/models` like you did with `/variantcache/logdog` so we don't have any potential interaction between the two crates' storage.

webern replied:
[afdc876](https://github.com/bottlerocket-os/bottlerocket/compare/6eb611d17506ab4aad98e5db4cd17b00bacd0839..afdc8765a9e97bb1d4436eda9b0c964fce24ab95)

tjkirch reviewed (commented):
I haven't been through the main Rust code yet, but these are my comments about the outlying bits.

webern commented:
> webern force-pushed the webern:logdog-ecs branch from 740b6eb to 6eb611d

Rebase to fix the Dockerfile conflict, no other changes yet.

webern commented:
> force-pushed the webern:logdog-ecs branch from 6eb611d to afdc876

Change the dockerfile and directory structure of the command files. Also fix a couple of smaller issues.

webern commented:
> webern force-pushed the webern:logdog-ecs branch from afdc876 to 7d1f985

Oops, create the ignored output directory if it does not exist.

bcressey commented on sources/logdog/conf/logdog.aws-ecs-1.conf:
```diff
@@ -0,0 +1,5 @@
+sh docker-daemon.json cat /etc/docker/daemon.json
+sh docker-info docker info
+sh ecs-agent-state.json cat /var/lib/ecs/data/ecs_agent_data.json
+sh ecs-config.json cat /etc/ecs/ecs.config.json
```
Can we do `exec` instead of `sh` since we don't have a shell?

webern replied:
[aa7d592](https://github.com/bottlerocket-os/bottlerocket/compare/60e5d3c8e65c58170e8f72d944a80be11ea098e0..aa7d59202dad5161a9dc8aff3995f1fbe0d4c706)

bcressey commented on sources/logdog/src/commands.rs:
```diff
@@ -0,0 +1,154 @@
+const VARIANT_COMMANDS: &str = std::include_str!("../conf/current/logdog.conf");
+
+/// Returns the list of commands to run by combining `COMMON_COMMANDS` and `VARIANT_COMMANDS`.
+/// These are read at compile-time from files named `common` and `variant` respectively.
+pub(crate) fn commands() -> Vec<&'static str> {
+    let mut commands: Vec<&str> = COMMON_COMMANDS
+        .lines()
+        .filter(|&command| !command.is_empty())
+        .collect();
+    let mut variant_commands: Vec<&str> = VARIANT_COMMANDS
+        .lines()
+        .filter(|&command| !command.is_empty())
+        .collect();
+    commands.append(&mut variant_commands);
+    commands
+}
```
You could also filter out lines that start with `#` so we could add comments for the commands explaining the relevance.

webern replied:
[aa7d592](https://github.com/bottlerocket-os/bottlerocket/compare/60e5d3c8e65c58170e8f72d944a80be11ea098e0..aa7d59202dad5161a9dc8aff3995f1fbe0d4c706)

tjkirch commented on Dockerfile:
```diff
@@ -35,7 +35,8 @@ ARG ARCH
 # We can't create directories via RUN in a scratch container, so take an existing one.
-COPY --chown=1000:1000 --from=sdk /tmp /variantcache
+COPY --chown=1000:1000 --from=sdk /tmp /variantcache/model
+COPY --chown=1000:1000 --from=sdk /tmp /variantcache/logdog
```
I know I suggested the subdirectories, but Ben told me that each usage of variantcache should actually be a new copy of the layer based on the earlier definition, and so subdirectories shouldn't actually be needed.  I'm not sure how to test that (`ls` in the models and logdog build scripts?) but it'd be cleaner not to use the subdirectories if not needed...

webern replied:
[aa7d592](https://github.com/bottlerocket-os/bottlerocket/compare/60e5d3c8e65c58170e8f72d944a80be11ea098e0..aa7d59202dad5161a9dc8aff3995f1fbe0d4c706)

tjkirch replied:
Did you find a way to check that the directories are actually empty at the start of the logdog and models build scripts?

webern replied:
I'll work on it this morning.

webern replied:
I added this to `logdog/build.rs`:

```rust
    if std::path::PathBuf::from("conf/current/models-foo").is_file() {
        panic!("models-foo exists!");
    }
    std::fs::write("conf/current/logdog-foo", "").unwrap();
```

And I added this to `models/build.rs`:

```rust
    if std::path::PathBuf::from("src/variant/logdog-foo").is_file() {
        panic!("logdog-foo exists!");
    }
    std::fs::write("src/variant/models-foo", "foo").unwrap();
```

`cargo make` succeeded, which I think shows that the two cache directories are independent.

tjkirch commented on sources/logdog/.gitignore:
```diff
@@ -0,0 +1,2 @@
+# ignore the symlink generated by build.rs
+conf/current/
```
```suggestion
/conf/current/
```
nit: It shouldn't actually make a difference here, but the starting slash at least makes clear to future readers that this directory lives at the top level.

webern replied:
[aa7d592](https://github.com/bottlerocket-os/bottlerocket/compare/60e5d3c8e65c58170e8f72d944a80be11ea098e0..aa7d59202dad5161a9dc8aff3995f1fbe0d4c706)

tjkirch commented on sources/logdog/build.rs:
```diff
@@ -30,3 +39,53 @@ fn main() {
+        format!("../{}", variant_filename),
+        "conf/current/logdog.conf",
+    )
+    .unwrap();
```
Can we give error messages the way `models` does rather than unwrapping?  I don't think the unwrap message will be very helpful, it usually doesn't include a path, and it won't explain why were trying to do this.

webern replied:
[aa7d592](https://github.com/bottlerocket-os/bottlerocket/compare/60e5d3c8e65c58170e8f72d944a80be11ea098e0..aa7d59202dad5161a9dc8aff3995f1fbe0d4c706)

tjkirch commented on sources/logdog/build.rs:
```diff
@@ -30,3 +39,53 @@ fn main() {
+        process::exit(1);
+    }
+    // create the conf/current directory if it is missing
+    fs::create_dir_all("conf/current").unwrap();
```
I think this might be simpler the way `models` does it, where there's an empty `.keep` file in the directory where we mount the writable cache (`conf/current` here) rather than needing to make the directory here.

webern replied:
[aa7d592](https://github.com/bottlerocket-os/bottlerocket/compare/60e5d3c8e65c58170e8f72d944a80be11ea098e0..aa7d59202dad5161a9dc8aff3995f1fbe0d4c706)

tjkirch commented on sources/logdog/src/error.rs:
```diff
@@ -3,92 +3,164 @@
     },
+
+    #[snafu(display("Encountered an empty command."))]
+    EmptyCommand { backtrace: Backtrace },
```
It seems this is unused now?

webern replied:
[aa7d592](https://github.com/bottlerocket-os/bottlerocket/compare/60e5d3c8e65c58170e8f72d944a80be11ea098e0..aa7d59202dad5161a9dc8aff3995f1fbe0d4c706)

tjkirch commented on sources/logdog/src/error.rs:
```diff
@@ -3,92 +3,164 @@
+    ConfigUtf8 {
+        source: std::str::Utf8Error,
+        path: PathBuf,
+    },
```
These seem unused?

webern replied:
[aa7d592](https://github.com/bottlerocket-os/bottlerocket/compare/60e5d3c8e65c58170e8f72d944a80be11ea098e0..aa7d59202dad5161a9dc8aff3995f1fbe0d4c706)

tjkirch commented on sources/logdog/src/main.rs:
```diff
@@ -18,22 +18,20 @@ For the commands used to gather logs, please see [log_request](src/log_request.r
-use log_request::{log_requests, LogRequest};
-use snafu::{ErrorCompat, OptionExt, ResultExt};
 use tempfile::TempDir;
+use From;
```
I've never seen something like this before, what is this doing?  Is it referencing `std::convert::From`?  It has to already be in scope to reference it like this, so I'm not sure what the additional `use` would do...

webern replied:
[aa7d592](https://github.com/bottlerocket-os/bottlerocket/compare/60e5d3c8e65c58170e8f72d944a80be11ea098e0..aa7d59202dad5161a9dc8aff3995f1fbe0d4c706)

tjkirch commented on sources/logdog/src/commands.rs:
```diff
@@ -0,0 +1,154 @@
+    P: AsRef<Path>,
+{
+    let request = request.as_ref();
+    let tokens = shell_words::split(request).context(error::CommandParse { command: request })?;
```
I'm not sure we want to split all command types using shell rules; it might not have the intended effect for other modes.  Can we do this inside `handle_shell_request` like we used to in `run_command`?

You could use `request.splitn(3, ' ')` to get an iterator of the mode, filename, and whatever's left, and use `.next().context(bla)` twice to set the mode and filename, and `.collect()` to get whatever's left into a string to pass to the handler, where `shell_words.split` (or whatever's appropriate for the mode) would come into play.

webern replied:
[aa7d592](https://github.com/bottlerocket-os/bottlerocket/compare/60e5d3c8e65c58170e8f72d944a80be11ea098e0..aa7d59202dad5161a9dc8aff3995f1fbe0d4c706)

tjkirch commented on sources/logdog/src/commands.rs:
```diff
@@ -0,0 +1,154 @@
+        if self.args.is_empty() {
+            format!("{} {}", self.mode, self.filename)
+        } else {
+            format!("{} {} {}", self.mode, self.filename, self.args.join(" "))
```
If we use shell_words to split, we should probably use it to join as well?

webern replied:
No longer applicable: [aa7d592](https://github.com/bottlerocket-os/bottlerocket/compare/60e5d3c8e65c58170e8f72d944a80be11ea098e0..aa7d59202dad5161a9dc8aff3995f1fbe0d4c706)

tjkirch commented on sources/logdog/src/commands.rs:
```diff
@@ -0,0 +1,154 @@
+/// hello world` and write the output to `hello.txt`. `http example http://example.com` will GET
+/// `http://example.com` and write the response body to `example`.
+#[derive(Debug, Clone)]
+struct Command<'a> {
```
```suggestion
struct LogRequest<'a> {
```
So that people used to `std::process::Command` don't get confused, and we don't need the full referencing of `std::process::Command` below.  And this was the terminology being used elsewhere :)

webern replied:
Scrubbed it all back to 'log request' terminology: [aa7d592](https://github.com/bottlerocket-os/bottlerocket/compare/60e5d3c8e65c58170e8f72d944a80be11ea098e0..aa7d59202dad5161a9dc8aff3995f1fbe0d4c706)

tjkirch commented on sources/logdog/src/commands.rs:
```diff
@@ -0,0 +1,154 @@
+        .filter(|&command| !command.is_empty())
+        .collect();
+    commands.append(&mut variant_commands);
+    commands
```
```suggestion
    COMMON_COMMANDS
        .lines()
        .chain(VARIANT_COMMANDS
        .lines())
        .filter(|&command| !command.is_empty() && !command.trim_start().starts_with('#'))
        .collect()
```
Shouldn't need to construct/expand two vecs and filter twice.

(Also fixes Ben's request)

webern replied:
[aa7d592](https://github.com/bottlerocket-os/bottlerocket/compare/60e5d3c8e65c58170e8f72d944a80be11ea098e0..aa7d59202dad5161a9dc8aff3995f1fbe0d4c706)

tjkirch commented on sources/logdog/src/commands.rs:
```diff
@@ -0,0 +1,154 @@
+        unmatched => {
+            return Err(error::Error::UnhandledRequest {
+                mode: unmatched.into(),
+                request: request.into(),
```
I don't think you need the manual `into` calls here; snafu should convert into owned strings if needed by the error variant, and only if there was a failure.

webern replied:
Not in this case since I am constructing the error myself. I'm fairly certain the into() calls will only execute when the code branch is executed.

tjkirch commented on sources/logdog/src/commands.rs:
```diff
@@ -0,0 +1,154 @@
+//!
+//! At build time, the `build.rs` file checks which variant is being built and creates a symlink
+//! file which points to the commands for the current variant. This file is named `variant`. We load
+//! the `variant` and `common` files into static strings at compile time, and these provide the list
```
I think these filenames are now out of date.  Same above `fn commands()`.

webern replied:
[aa7d592](https://github.com/bottlerocket-os/bottlerocket/compare/60e5d3c8e65c58170e8f72d944a80be11ea098e0..aa7d59202dad5161a9dc8aff3995f1fbe0d4c706)

webern commented:
> webern force-pushed the webern:logdog-ecs branch from 7d1f985 to 60e5d3c

Rebase develop without any other changes.

webern committed:
logdog: order and format errors

webern committed:
logdog: add support for ecs

webern commented:
> webern force-pushed the webern:logdog-ecs branch from 60e5d3c to aa7d592 

Un-renamed `log_request.rs` / `command.rs`. Re-ordered the error re-ordering commit to happen first. Addressed all PR comments.

webern commented:
> webern force-pushed the webern:logdog-ecs branch from aa7d592 to cb6b585

A fix in the readme.

tjkirch commented on sources/logdog/src/main.rs:
```diff
@@ -12,7 +12,12 @@ logs are at: /tmp/bottlerocket-logs.tar.gz
-For the commands used to gather logs, please see [log_request](src/log_request.rs).
+For the log requests used to gather logs, please see the following:
+
+* [log_requests](src/log_requests.rs)
```
Looks like the last update might have fixed the README directly but not the source here.

tjkirch commented on sources/logdog/build.rs:
```diff
@@ -30,3 +89,9 @@ fn main() {
 }
+
+fn main() {
+    println!("cargo:rerun-if-env-changed={}", ENV_VARIANT);
+    symlink_variant();
```
Can we do the print inside the function so it's clearer why it's relevant?

tjkirch commented on sources/logdog/src/log_request.rs:
```diff
@@ -1,29 +1,238 @@
+        if self.instructions.is_empty() {
+            format!("{} {}", self.mode, self.filename)
+        } else {
+            format!("{} {} {}", self.mode, self.filename, self.instructions,)
```
```suggestion
            format!("{} {} {}", self.mode, self.filename, self.instructions)
```

webern replied:
A bit disappointed that fmt didn't fix this.

webern commented:
> webern force-pushed the webern:logdog-ecs branch from cb6b585 to 3ed71ad

I missed this readme change in the previous push.

tjkirch commented on sources/logdog/src/log_request.rs:
```diff
@@ -1,29 +1,238 @@
+/// The `logdog` log requests that all variants have in common.
+const COMMON_REQUESTS: &str = std::include_str!("../conf/logdog.common.conf");
+/// The `logdog` log requests that are specific to the current variant.
+const VARIANT_REQUESTS: &str = std::include_str!("../conf/current/logdog.conf");
```
```suggestion
const COMMON_REQUESTS: &str = include_str!("../conf/logdog.common.conf");
/// The `logdog` log requests that are specific to the current variant.
const VARIANT_REQUESTS: &str = include_str!("../conf/current/logdog.conf");
```

tjkirch commented on sources/logdog/src/log_request.rs:
```diff
@@ -1,29 +1,238 @@
+    let stderr_file = ofile
+        .try_clone()
+        .context(error::CommandErrFile { path: &outpath })?;
+    std::process::Command::new(command)
```
Can we `use` this instead of using the full path now that there's no naming conflict?

tjkirch commented on sources/logdog/src/log_request.rs:
```diff
@@ -1,29 +1,238 @@
+    let data = response.bytes().context(error::HttpResponseBytes {
+        request: request.to_string(),
+    })?;
+    std::fs::write(&outpath, &data).context(error::HttpWriteBytes {
```
```suggestion
    fs::write(&outpath, &data).context(error::HttpWriteBytes {
```

tjkirch commented on sources/logdog/src/log_request.rs:
```diff
@@ -1,29 +1,238 @@
+    );
+    let dest = tempdir.as_ref().join(request.filename);
+    let _ = fs::copy(&request.instructions, &dest).context(error::FileCopy {
+        request: request.to_string(),
```
Do you need the manual `to_string` here?  If so, can we do a `with_context`?

webern replied:
It is needed, so switched to with_context.

tjkirch replied:
Same thing on the line above?

webern commented:
> webern force-pushed the webern:logdog-ecs branch from 3ed71ad to f8f85c6

Final fixes and cleanups.

webern commented:
>  webern force-pushed the webern:logdog-ecs branch from f8f85c6 to ab8692e

Avoid allocations when not-needed by using `with_context`.

webern commented:
> webern force-pushed the webern:logdog-ecs branch from ab8692e to a01449d 

File copy doesn't work for /proc/mounts so using cat command instead. Also /etc/mtab (which also can not be copied with a file request) is duplicative of /proc/mounts, so removed it.
